HW1/P4/P4.py

```diff
@@ -0,0 +1,50 @@
+# Your code here
+import findspark
+findspark.init()
+
+import pyspark
+sc = pyspark.SparkContext(appName="Spark1")
+
+from P4_bfs import *
+
+# Reduce the Spark Logs
+logger = sc._jvm.org.apache.log4j
+logger.LogManager.getLogger("org"). setLevel( logger.Level.ERROR )
+logger.LogManager.getLogger("akka").setLevel( logger.Level.ERROR )
+
+import numpy as np 
+import itertools
+
+def remove_quotes(s):
+    return s.replace('"','')
+
+def get_pairs(arr):
+    return list(itertools.permutations(arr, 2))
+
+num_Partitions = 16
+
+#Extract the strings and remove all unnessary quotes
+wlist = sc.textFile('source.csv', num_Partitions)
+character_to_book = wlist.map(lambda x: x.split('","')).map(lambda x: map(remove_quotes, x))
+
+#Switch the order of the map to be from book to characters
+book_to_character = character_to_book.map(lambda (x, y): (y, x))
+book_to_characters = book_to_character.groupByKey(numPartitions = num_Partitions)
+
+# Extract all pairs of characters that share a book.
+adjacent_heros = book_to_characters.flatMap(lambda (book, characters): get_pairs(characters)).distinct().partitionBy(num_Partitions)
+
+# Create the graph as an adjacency list.
+graph = adjacent_heros.groupByKey().partitionBy(num_Partitions).cache()
+
```

Comment:
Your code is very cleanly laid out here, and it's easy to follow the steps in creating and modifying the graph